flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperator.java

```diff
@@ -136,19 +154,28 @@ private void advanceWatermark() {
 
     @Override
     public void onProcessingTime(long timestamp) throws Exception {
-        advanceWatermark();
+        // timestamp and now can be off in case TM is heavily overloaded.
+        long now = getProcessingTimeService().getCurrentProcessingTime();
 
-        if (idleTimeout > 0 && currentStatus.equals(WatermarkStatus.ACTIVE)) {
-            final long currentTime = getProcessingTimeService().getCurrentProcessingTime();
-            if (currentTime - lastRecordTime > idleTimeout) {
-                // mark the channel as idle to ignore watermarks from this channel
-                emitWatermarkStatus(WatermarkStatus.IDLE);
-            }
+        if (watermarkInterval > 0
+                && lastWatermarkPeriodicEmitTime + watermarkInterval <= timestamp) {
```

Comment:
As the comment said, `timestamp and now` may be different. I'm curious, why using `timestamp` instead  of `now` here?

If the timer is triggered later, and `lastWatermarkPeriodicEmitTime + watermarkInterval` may be less than `now`, but greater than `timestamp`. For this case, we could `advanceWatermark`, right?

It means we can use `now` directly. For old code, it uses `currentTime` as well.

Reply:
Goog point I think. I will try to switch to `now` and let's see if some tests will complain or not.